clippy_utils/src/lib.rs

```diff
@@ -517,18 +518,33 @@ pub fn path_to_res(cx: &LateContext<'_>, path: &[&str]) -> Res {
             .iter()
             .find(|item| item.ident.name.as_str() == name)
     }
+    fn primitive_impl<'tcx>(tcx: TyCtxt<'tcx>, base: &str, first: &str) -> Option<&'tcx Export> {
+        if let Some(&(index, Target::Impl)) = lang_items::ITEM_REFS.get(&Symbol::intern(base)) {
+            let def_id: DefId = tcx.lang_items().items()[index]?;
 
-    let (krate, first, path) = match *path {
-        [krate, first, ref path @ ..] => (krate, first, path),
+            item_child_by_name(tcx, def_id, first)
+        } else {
+            None
+        }
+    }
+
+    let (base, first, path) = match *path {
+        [base, first, ref path @ ..] => (base, first, path),
         [primitive] => {
             return PrimTy::from_name(Symbol::intern(primitive)).map_or(Res::Err, Res::PrimTy);
         },
         _ => return Res::Err,
     };
     let tcx = cx.tcx;
     let crates = tcx.crates(());
-    let krate = try_res!(crates.iter().find(|&&num| tcx.crate_name(num).as_str() == krate));
-    let first = try_res!(item_child_by_name(tcx, krate.as_def_id(), first));
+    let first = try_res!(
+        crates
+            .iter()
+            .find(|&&num| tcx.crate_name(num).as_str() == base)
+            .and_then(|krate| item_child_by_name(tcx, krate.as_def_id(), first))
+            .or_else(|| primitive_impl(tcx, base, first))
```

Comment:
I think the primitive check should come before iterating the crates, and the crate/primitive cases can collapse into the same code path before calling `item_child_by_name`.